### PR TITLE
[exporter/loadbalancing] Fix memory leaks

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

This fixes a few goroutine leaks in the loadbalancing exporter.

1. `metrics`, `traces`, and `logs` exporters were starting their respective load balancers, but were not shutting them down. This adds each respective shutdown call.
2. The `loadbalancer` was starting the resolver but never shutting it down. This adds a shutdown call to the resolver.
3. The static resolver was starting resolvers for each passed in exporter, but never shut them down. This adds a shutdown call for each resolver in the static resolver.

Also added a couple missing `Shutdown` calls from tests.

**Link to tracking Issue:** <Issue number if applicable>
#30438

**Testing:** <Describe what testing was performed and which tests were added.>
All existing tests are passing as well as added goleak checks.

### CODE REVIEW

File: exporter/loadbalancingexporter/resolver_static.go [hunk 1]
```diff
@@ -49,13 +50,23 @@ func (r *staticResolver) start(ctx context.Context) error {
...
 
-func (r *staticResolver) shutdown(_ context.Context) error {
+func (r *staticResolver) shutdown(ctx context.Context) error {
+	r.endpoints = nil
+
+	r.mx.Lock()
+	defer r.mx.Unlock()
+	for _, callback := range r.onChangeCallbacks {
+		callback(r.endpoints)
+	}
```
Comment (crobert-1) [inline exporter/loadbalancingexporter/resolver_static.go:60]: This solution could use some discussion or feedback. The resolver was [specifically designed to only be called once](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1412#discussion_r518319655), as it's for a static list backends that shouldn't change.

However, the resolver [starts each passed in exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e2d601acb20fe0fa1c99d860ff484a208341af27/exporter/loadbalancingexporter/loadbalancer.go#L136), and due to the `once.Do` functionality, will never shutdown these started exporters. To use existing functionality as much as possible, if we clear the static resolver's list of endpoints and call the `onChangeCallbacks` methods, the existing exporters will be shutdown.

I added the mutex to make sure we're not calling `resolve` concurrently, but there may be a cleaner way to do this. The requirement here is that `shutdown` is called after the initial `start` method calls `resolve`.

Comment (crobert-1) [reply exporter/loadbalancingexporter/resolver_static.go:60]: I can file an issue for this if it's preferable to discuss elsewhere, just let me know.

Comment (jpkrohling) [reply exporter/loadbalancingexporter/resolver_static.go:60]: I think the approach here is fine, thank you for catching this!

**Review by jpkrohling — APPROVED** [review]

File: exporter/loadbalancingexporter/resolver_static.go (hunk 1 above)
Comment (dmitryax) [reply exporter/loadbalancingexporter/resolver_static.go:60]: Why do we need an extra lock here? I see `resolve` is called synchronously at the start of the load balancer. I don't think we allow calling Start and Shutdown concurrently.

Comment (crobert-1) [reply exporter/loadbalancingexporter/resolver_static.go:60]: I believe this was simply done to be safe. If it's more confusing than helpful I can remove it.

Comment (dmitryax) [reply exporter/loadbalancingexporter/resolver_static.go:60]: I would prefer not introducing locks that are not needed

Comment (crobert-1) [reply exporter/loadbalancingexporter/resolver_static.go]: I've removed the mutex.

**Comment by crobert-1** [comment]
CI/CD failures are #31494 and #31443

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**